sunpy/map/mapbase.py

```diff
@@ -1978,50 +1978,130 @@ def draw_grid(self, axes=None, grid_spacing: u.deg = 15*u.deg, annotate=True, **
-            radius = self.rsun_obs.to(u.deg).value
-        else:
-            radius = self.rsun_obs.value
-        c_kw = {'radius': radius,
-                'fill': False,
+        c_kw = {'fill': False,
                 'color': 'white',
-                'zorder': 100,
-                'transform': transform
-                }
+                'zorder': 100}
         c_kw.update(kwargs)
 
-        circ = patches.Circle([0, 0], **c_kw)
-        axes.add_artist(circ)
-
-        return [circ]
+        # Obtain the solar radius and the world->pixel transform
+        if not is_wcsaxes:
+            radius = self.rsun_obs.value
+            transform = axes.transData
+        else:
+            radius = self.rsun_obs.to_value(u.deg)
+            transform = axes.get_transform('world')
+
+        # transform is always passed on as a keyword argument
+        c_kw.setdefault('transform', transform)
+
+        # If not WCSAxes or if the map's frame matches the axes's frame and is Helioprojective,
+        # we can use Circle
+        if not is_wcsaxes or (self.coordinate_frame == axes._transform_pixel2world.frame_out
+                              and isinstance(self.coordinate_frame, Helioprojective)):
+            c_kw.setdefault('radius', radius)
+
+            circ = patches.Circle([0, 0], **c_kw)
+            axes.add_artist(circ)
+
+            return circ, None
+
+        # Otherwise, we use Polygon to be able to distort the limb
+
+        # Create the limb coordinate array using Heliocentric Radial
+        limb_radial_distance = self.observer_coordinate.radius * np.cos(self.rsun_obs)
+        limb_hcr_rho = limb_radial_distance * np.sin(self.rsun_obs)
+        limb_hcr_z = self.observer_coordinate.radius - limb_radial_distance * np.cos(self.rsun_obs)
+        limb_hcr_psi = np.linspace(0, 2*np.pi, resolution+1)[:-1] << u.rad
+        limb = SkyCoord(limb_hcr_rho, limb_hcr_psi, limb_hcr_z, representation_type='cylindrical',
+                        frame='heliocentric', observer=self.observer_coordinate, obstime=self.date)
+
+        # Transform the limb to the axes frame and get the 2D vertices
+        axes_frame = axes._transform_pixel2world.frame_out
+        limb_in_axes = limb.transform_to(axes_frame)
+        Tx = limb_in_axes.spherical.lon.to_value(u.deg)
+        Ty = limb_in_axes.spherical.lat.to_value(u.deg)
+        vertices = np.array([Tx, Ty]).T
+
+        # Determine which points are visible
+        if hasattr(axes_frame, 'observer'):
+            # The reference distance is the distance to the limb for the axes observer
+            rsun = getattr(axes_frame, 'rsun', self.rsun_meters)
+            reference_distance = np.sqrt(axes_frame.observer.radius**2 - rsun**2)
+            is_visible = limb_in_axes.spherical.distance <= reference_distance
+        else:
+            # If the axes has no observer, the entire limb is considered visible
+            is_visible = np.ones_like(limb_in_axes.spherical.distance, np.bool)
```

Comment:
```suggestion
            is_visible = np.ones_like(limb_in_axes.spherical.distance.value, np.bool)
```
I had to make this small change to make this PR work on a Carrington map. Without it `is_visible` is a boolean array with dimensions `'m``, so the `~is_visible` call below errors out.

Reply:
Thanks for catching this bug.  I should have included `subok=False`, which should be slightly better performing that your suggestion.

However, I'm a little confused.  If you're using Carrington coordinates for your WCSAxes, it shouldn't even fall into this code branch.  Are you using something other than Carrington coordinates?

Reply:
I think it's because I manually created a header without observer information using the `header_helper`. Full code is [here](https://solar-orbiter-python.readthedocs.io/en/latest/_examples_build/remote_sensing/eui_and_aia.html#sphx-glr-examples-build-remote-sensing-eui-and-aia-py) if it's helpful.

Reply:
Ah, yes, your WCS is in Stonyhurst heliographic coordinates, not in Carrington heliographic coordinates, so no observer.

Reply:
Added a figure test